### DISCUSSION TITLE
Upgrading Firefox cask to v113 somehow corrupted the cask binary

### DISCUSSION THREAD

**Discussion by fabianschulz** [Homebrew/discussions#4498] [Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.17
ORIGIN: https://github.com/Homebrew/brew
HEAD: c0f8068573bbf886fdc78d3f9daca54d53e7e49e
Last commit: 2 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 1d8c58a0aad7953501bd69df9f5014a6dfbf199e
Core tap last commit: 3 months ago
Core tap branch: master
Core tap JSON: 10 May 15:27 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: You have an unnecessary local Cask tap.
This can cause problems installing up-to-date casks.
Please remove it by running:
  brew untap homebrew/cask

Warning: You have an unnecessary local Core tap!
This can cause problems installing up-to-date formulae.
Please remove it by running:
 brew untap homebrew/core
```

### Description of issue

Hi, I wanted to update my Firefox browser to the latest version, but somehow the upgrade process corrupted the installation and I'm unsure what to do next. I am afraid of removing and re-installing Firefox from scratch, since I have a ton of privacy addons configured. 

I ran this as usual:

fabian@MacBook ~ % brew upgrade --greedy 
==> Upgrading 1 outdated package:
firefox 112.0.2 -> 113.0

But for some reason something went wrong during installation of the v113 binary of Firefox:

==> Upgrading 1 outdated package:
firefox 112.0.2 -> 113.0
==> Upgrading firefox
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-cask/(...)/Casks/firefox.rb
...100.0%
==> Downloading https://download-installer.cdn.mozilla.net/pub/firefox/releases/113.0/mac/en-US/Firefox%20113.0.dmg
...100.0%
==> Backing App 'Firefox.app' up to '/opt/homebrew/Caskroom/firefox/112.0.2/Firefox.app'
==> Removing App '/Applications/Firefox.app'
==> Unlinking Binary '/opt/homebrew/bin/firefox'
==> Moving App 'Firefox.app' to '/Applications/Firefox.app'
==> Purging files for version 113.0 of Cask firefox
==> Purging files for version 113.0 of Cask firefox
Warning: Reverting upgrade for Cask firefox
==> Purging files for version 112.0.2 of Cask firefox
Error: firefox: undefined method `artifacts' for #<Hash:0x0000000140ac9658>
Please report this issue:
  https://docs.brew.sh/Troubleshooting
/opt/homebrew/Library/Homebrew/cask/upgrade.rb:127:in upgrade_casks
/opt/homebrew/Library/Homebrew/cmd/upgrade.rb:253:in upgrade_outdated_casks
/opt/homebrew/Library/Homebrew/cmd/upgrade.rb:137:in upgrade
/opt/homebrew/Library/Homebrew/brew.rb:94:in main 

When trying to launch Firefox, Finder now complains that it "may be damaged or incomplete" and it cannot be started (writing this post on Safari).

Is there any way I can "resume" the upgrade?

Many thanks!
_[Edit: I had to remove some characters like ' and <> in the GitHub comment field since it messed up the formatting]_

### Relevant casks

Firefox

**Comment by timfallmk** [comment]
As someone else noted, this is likely related to changes in the upgrade mechanism of casks (to prevent them being quarantined/removed from the dock). I had a similar problem with the most recent upgrade of chrome and chrome canary.
https://github.com/Homebrew/brew/pull/15138

**Reply by fabianschulz** [reply]
Thanks for pointing this out @timfallmk ! If there is anything I can help with testing, I am happy to do so. If you had the same issue with your Chrome cask: did you find a way to "fix" the (broken) upgrade? If it's a bug in Brew itself and there is a hotfix in sight, I will of course be patient.

**Reply by timfallmk** [reply, marked as answer]
Forcing a reinstall of the affected casks got me back on the road with no loss of settings.
```shell
brew install --cask google-chrome --force
```

**Reply by fabianschulz** [reply]
> Forcing a reinstall of the affected casks got me back on the road with no loss of settings.
> 
> ```shell
> brew install --cask google-chrome --force
> ```

Thank you @timfallmk , I can confirm that this "fixed" the broken cask upgrade. Settings and addons were fortunately not affected. I'll mark the thread as answered, but of course the issue in https://github.com/Homebrew/brew/pull/15138 still persists until the root cause will be found.

**Comment by timfallmk** [comment]
Adding similar behavior from `google-chrome` and `google-chrome-canary`
```shell
==> Upgrading google-chrome to 113.0.5672.92
==> Downloading https://dl.google.com/chrome/mac/universal/stable/GGRO/googlechrome.dmg
######################################################################################################################################################################################################################################################### 100.0%
Warning: No checksum defined for cask 'google-chrome', skipping verification.
==> Uninstalling Cask google-chrome
==> Backing App 'Google Chrome.app' up to '/opt/homebrew/Caskroom/google-chrome/113.0.5672.63/Google Chrome.app'
==> Removing App '/Applications/Google Chrome.app'
==> Purging files for version 113.0.5672.63 of Cask google-chrome
==> Installing Cask google-chrome
==> Moving App 'Google Chrome.app' to '/Applications/Google Chrome.app'
cp: /opt/homebrew/Caskroom/google-chrome/113.0.5672.92/Google Chrome.app/*: No such file or directory
cp: /opt/homebrew/Caskroom/google-chrome/113.0.5672.92/Google Chrome.app/.*: No such file or directory
==> Purging files for version 113.0.5672.92 of Cask google-chrome
Error: Failure while executing; `/usr/bin/sudo -E -- /bin/cp -pR /opt/homebrew/Caskroom/google-chrome/113.0.5672.92/Google\ Chrome.app/\* /opt/homebrew/Caskroom/google-chrome/113.0.5672.92/Google\ Chrome.app/.\* /Applications/Google\ Chrome.app/` exited with 1. Here's the output:
cp: /opt/homebrew/Caskroom/google-chrome/113.0.5672.92/Google Chrome.app/*: No such file or directory
cp: /opt/homebrew/Caskroom/google-chrome/113.0.5672.92/Google Chrome.app/.*: No such file or directory

==> Upgrading google-chrome-canary to 115.0.5759.0
==> Downloading https://dl.google.com/chrome/mac/universal/canary/googlechromecanary.dmg
######################################################################################################################################################################################################################################################### 100.0%
Warning: No checksum defined for cask 'google-chrome-canary', skipping verification.
==> Uninstalling Cask google-chrome-canary
==> Backing App 'Google Chrome Canary.app' up to '/opt/homebrew/Caskroom/google-chrome-canary/115.0.5751.0/Google Chrome Canary.app'
==> Removing App '/Applications/Google Chrome Canary.app'
==> Purging files for version 115.0.5751.0 of Cask google-chrome-canary
==> Installing Cask google-chrome-canary
==> Moving App 'Google Chrome Canary.app' to '/Applications/Google Chrome Canary.app'
cp: /opt/homebrew/Caskroom/google-chrome-canary/115.0.5759.0/Google Chrome Canary.app/*: No such file or directory
cp: /opt/homebrew/Caskroom/google-chrome-canary/115.0.5759.0/Google Chrome Canary.app/.*: No such file or directory
==> Purging files for version 115.0.5759.0 of Cask google-chrome-canary
Error: Failure while executing; `/usr/bin/sudo -E -- /bin/cp -pR /opt/homebrew/Caskroom/google-chrome-canary/115.0.5759.0/Google\ Chrome\ Canary.app/\* /opt/homebrew/Caskroom/google-chrome-canary/115.0.5759.0/Google\ Chrome\ Canary.app/.\* /Applications/Google\ Chrome\ Canary.app/` exited with 1. Here's the output:
cp: /opt/homebrew/Caskroom/google-chrome-canary/115.0.5759.0/Google Chrome Canary.app/*: No such file or directory
cp: /opt/homebrew/Caskroom/google-chrome-canary/115.0.5759.0/Google Chrome Canary.app/.*: No such file or directory
```

**Comment by JBYoshi** [comment]
I'm able to consistently reproduce this using `brew reinstall firefox`. I'm currently trying to diagnose it.

**Reply by JBYoshi** [reply]
I haven't gotten the "undefined method: artifacts" error, but I've traced the "cp: [...].app/*: No such file or directory" error down to the wildcards in the `cp` command not being expanded.

**Reply by JBYoshi** [reply]
https://github.com/Homebrew/brew/pull/15407 is up to handle the wildcard problem.

**Reply by fabianschulz** [reply]
Thank you so much for digging into this @JBYoshi - really appreciate your help!

**Comment by JBYoshi** [comment]
Someone else is reporting the same "undefined method: artifacts" error in https://github.com/Homebrew/brew/issues/15406. I'm still not sure how to reproduce that specific error, so if you have some consistent steps to do that please let me know.

**Reply by JBYoshi** [reply]
I've tracked down the direct issue through some extra work with the type checker, and I'll submit a fix PR shortly once I add some unit tests for this case. However, the reason it wasn't caught earlier was because the relevant part of the code only runs as part of error handling recovery, which implies that there's another error going on too that's getting suppressed. Hopefully, once this particular error is fixed, we'll be able to track down the root cause.

**Reply by JBYoshi** [reply]
https://github.com/Homebrew/brew/pull/15412 has been submitted to fix the "undefined method: artifacts" error.

**Reply by JBYoshi** [reply]
That PR has been merged, so that particular error should now be fixed.

**Comment by fabianschulz** [comment]
What I find strange is that I ran the exact same thing today (with the minor release of Firefox 113.0.1) and this time it went through without problems:

fabian@MacBook ~ % brew upgrade --greedy                                                                       
==> Upgrading 1 outdated package:
firefox 113.0 -> 113.0.1
==> Upgrading firefox
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-cask/(...)/Casks/firefox.rb
100.0%
==> Downloading https://download-installer.cdn.mozilla.net/pub/firefox/releases/113.0.1/mac/en-US/Firefox%20113.0.1.dmg
100.0%
==> Backing App 'Firefox.app' up to '/opt/homebrew/Caskroom/firefox/113.0/Firefox.app'
==> Removing App '/Applications/Firefox.app'
==> Unlinking Binary '/opt/homebrew/bin/firefox'
==> Moving App 'Firefox.app' to '/Applications/Firefox.app'
==> Linking Binary 'firefox.wrapper.sh' to '/opt/homebrew/bin/firefox'
==> Purging files for version 113.0 of Cask firefox
🍺  firefox was successfully upgraded!

Any idea why this time it worked as expected? AFAIK nothing was changed on Brew side or was committed since I reported the issue. I am curious if it was just me this time or if this issue still persists when upgrading other casks (such as Chrome, as mentioned by @timfallmk).

**Reply by SMillerDev** [reply]
See https://github.com/orgs/Homebrew/discussions/4498#discussioncomment-5891976, it was fixed

**Reply by JBYoshi** [reply]
There hasn't been a stable release with these fixes yet, so unless you're in developer mode or following the latest commits, the fix won't have been rolled out yet.

**Reply by fabianschulz** [reply]
Thank you @JBYoshi & @SMillerDev for taking care of that bug. Much appreciated!
